### ISSUE TITLE
System.InvalidCastException: Specified cast is not valid. at Xamarin.Auth.WebAuthenticatorController+UIWebViewDelegate.ShouldStartLoad

### ISSUE THREAD

**Issue by JakubHolovsky** [xamarin/Xamarin.Auth#164]
Using the latest nuget package 1.4.1 I get following exception when trying to show up the Twitter login.

The exception comes up at this line:

```
static void Main (string[] args)
		{
			UIApplication.Main (args, typeof(Application), typeof(AppDelegate));
		}
```

This is how I am launching the view controller for authenticator:

```
if (authenticator.IsUsingNativeUI)
		    {
		        controller.ActivateController((SFSafariViewController)authenticator.GetUI());
		    }
		    else
		    {
		        controller.ActivateController((UIViewController)authenticator.GetUI());
		    }
```

```
System.InvalidCastException: Specified cast is not valid.
  at Xamarin.Auth.WebAuthenticatorController+UIWebViewDelegate.ShouldStartLoad (UIKit.UIWebView webView, Foundation.NSUrlRequest request, UIKit.UIWebViewNavigationType navigationType) [0x0000b] in <aa21ef55360b447cbbfb95f4ba40f14f>:0 
  at (wrapper managed-to-native) UIKit.UIApplication:UIApplicationMain (int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Users/builder/data/lanes/4466/a04678c2/source/xamarin-macios/src/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x00038] in /Users/builder/data/lanes/4466/a04678c2/source/xamarin-macios/src/UIKit/UIApplication.cs:63 
  at UIKit.UIApplication.Main (System.String[] args, System.Type principalClass, System.Type delegateClass) [0x00
000] in /Users/builder/data/lanes/4466/a04678c2/source/xamarin-macios/src/UIKit/UIApplication.cs:68 
  at Psonar.Apps.iOS.PayPerPlay.Program.Main (System.String[] args) [0x00001] in E:\Dev\psonar\Source\Psonar.Apps\Psonar.Apps.iOS\Psonar.Apps.iOS.PayPerPlay\Main.cs:9 
2017-04-06 15:39:21.878 PsonarAppsiOSPayPerPlay[18840:239335] Unhandled managed exception:
Specified cast is not valid. (System.InvalidCastException)
  at Xamarin.Auth.WebAuthenticatorController+UIWebViewDelegate.ShouldStartLoad (UIKit.UIWebView webView, Foundation.NSUrlRequest request, UIKit.UIWebViewNavigationType navigationType) [0x0000b] in <aa21ef55360b447cbbfb95f4ba40f14f>:0 
  at (wrapper managed-to-native) UIKit.UIApplication:UIApplicationMain (int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Users/builder/data/lanes/4466/a04678c2/source/xamarin-macios/src/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x00038] in /Users/builder/data/lanes/4466/a04678c2/source/xamarin-macios/src/UIKit/UIApplication.cs:63 
  at UIKit.UIApplication.Main (System.String[] args, System.Type principalClass, System.Type delegateClas
s) [0x00000] in /Users/builder/data/lanes/4466/a04678c2/source/xamarin-macios/src/UIKit/UIApplication.cs:68 
  at Psonar.Apps.iOS.PayPerPlay.Program.Main (System.String[] args) [0x00001] in E:\Dev\psonar\Source\Psonar.Apps\Psonar.Apps.iOS\Psonar.Apps.iOS.PayPerPlay\Main.cs:9 
2017-04-06 15:39:21.878 PsonarAppsiOSPayPerPlay[18840:239335] critical: Stacktrace:
2017-04-06 15:39:21.879 PsonarAppsiOSPayPerPlay[18840:239335] critical: 
Native stacktrace:
2017-04-06 15:39:21.902 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	0   PsonarAppsiOSPayPerPlay             0x000000010f6c45fd mono_handle_native_sigsegv + 253
2017-04-06 15:39:21.902 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	1   libsystem_platform.dylib            0x000000011c7f8b3a _sigtramp + 26
2017-04-06 15:39:21.903 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	2   ???                                 0x000000010feac0e0 0x0 + 4562010336
2017-04-06 15:39:21.903 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	3   libsystem_c.dylib                   0x000000011c56788f abort + 127
2017-04-06 15:39:21.903 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	4   PsonarAppsiOSPayPerPlay             0x000000010f86c98f xamarin_unhandled_exception_handler + 47
2017-04-06 15:39:21.904 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	5   PsonarAppsiOSPayPerPlay             0x000000010f723c54 mono_invoke_unhandled_exception_hook + 148
2017-04-06 15:39:21.904 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	6   PsonarAppsiOSPayPerPlay             0x000000010f6c40e2 mono_handle_exception_internal + 4626
2017-04-06 15:39:21.905 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	7   PsonarAppsiOSPayPerPlay             0x000000010f6c2ec8 mono_handle_exception + 24
2017-04-06 15:39:21.905 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	8   PsonarAppsiOSPayPerPlay             0x000000010f63cc8f mono_amd64_throw_exception + 143
2017-04-06 15:39:21.905 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	9   PsonarAppsiOSPayPerPlay             0x000000010f63ccf6 mono_amd64_throw_corlib_exception + 70
2017-04-06 15:39:21.906 PsonarAppsiOSPayPerPlay[18840:239335] critical: 	10  ???                                 0x000000012ff97ee3 0x0 + 5099847395
2017-04-06 15:39:21.906 PsonarAppsiOSPayPerPlay[18840:239335] critical: 
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================
The app has been terminated.
Launch failed. The app 'Psonar.Apps.iOS.PayPerPlay' could not be launched on 'iPhone 5s iOS 10.3'. Error: An error occurred while executing MTouch. Please check the logs for more details.
The app has been terminated.

```

**Comment by JakubHolovsky** [comment]
I wonder if this can be coming from this cast but I am not too sure:

```
WebRedirectAuthenticator wra = null;
wra = (WebRedirectAuthenticator)this.controller.authenticator;
```

This is in the ShouldStartLoad method.

**Comment by JakubHolovsky** [comment]
Does not happen on older version (much older version).

**Comment by nickmathewsPerf** [comment]
I'm getting this same problem on iOS, but not on Android. Have you figured out any solutions?

**Comment by JakubHolovsky** [comment]
@nickmathewsPerf haven't figured out any, this is specific to iOS only. 

I guess I could test it with the source code of Xamarin.Auth but seems like a bit of hassle. So I might leave that to the official Xamarin.Auth team. My solution for now was to use older version of the library...

**Comment by anarchicknight** [comment]
I'm also experiencing this issue on 1.4.1.0

**Comment by moljac** [comment]
@JakubHolosvsky @anarchicknight 
It is the cast for sure iOS only - in UIWebViewDelegate parsing newly added URLs like (http://localhost, http://127.0.0.1, http://::1) and custom schemes, but what I cannot reproduce that at all. Namely both OAuth2Authenticator and OAuth1Authenticator are WebRedirectAuthenticators, so cast should be OK. 
Are you using something else to authenticate?
Is this Forms app?
thanks

**Comment by JakubHolovsky** [comment]
@moljac we don't use anything else to authenticate, just Xamarin.Auth. It is Xamarin.iOS app.

**Comment by moljac** [comment]
@JakubHolovsky 
I meant something other in Xamarin.Auth, but OAuth2Authenticator (or OAuth1Authenticator), namely this cast is in the part where redirect_url is verified and parsed. I added support for localhost (which was not in Xamarin.Auth before). 
The cast is to WebRedirectAuthenticator and OAuth1 and OAuth2 use redirecting flow, so I thought you might have used some other Authenticator in the Xamarin.Auth.

Can you give me more info about redirect_url? if possible.

**Comment by JakubHolovsky** [comment]
@moljac using only OAuth1Authenticator. This is the way how we initialize it:

```
var authenticator = new OAuth1Authenticator(Platform.TwitterConsumerKey, Platform.TwitterConsumerSecret,
				new Uri("https://api.twitter.com/oauth/request_token"),
				new Uri("https://api.twitter.com/oauth/authorize"),
				new Uri("https://api.twitter.com/oauth/access_token"), new Uri("https://mobile.twitter.com/home"));
```

**Comment by moljac** [comment]
@JakubHolovsky 
Looks OK. 
Which nuget? Can you try 1.5.0-alpha and give me some feedback pls?

**Comment by JakubHolovsky** [comment]
@moljac tried with the latest alpha, not getting a crash anymore but doesn't load the twitter page at all. Basically shows empty Authenticate view.

**Comment by moljac** [comment]
@JakubHolovsky 
Please I need more info. The whole library got so big, not only platform coverage but features too. CustomTabs alone is story for itself (diffs in API, nugets, Chrome installed).

Honestly I didn't test/tweak providers (OAuth providers = test cases for the lib) for quite some time. I heard Twitter now supports OAuth2 (I think I saw it when I was on their dev portal), so they have changed something for sure. I will check FB and Twitter ASAP.

**Comment by JakubHolovsky** [comment]
@moljac I am testing this on simulator so it's reasonably restricted / nothing extra installed.

**Comment by anarchicknight** [comment]
Unfortunately I can not provide any more info as I have stopped using Xamarin for my project

**Comment by moljac** [comment]
Fixed in 1.5.0
Closing
